### PR TITLE
feat!: add compute plan builder

### PR DESCRIPTION
**Pull request by ThibaultFy**
## Summary
TO BE REVIEW COMMIT BY COMMIT

First commit:
- move the schemas file into the strategy folder. This schema file is used to implement the scheme of the different Shared stated of each strategy. With the introduction of the compute plan builder, it does not makes sense to have the file at the root of the lib.

Second commit:
- Create a ComputePlanBuilder object to be inherited for every experiment to be run with substrafl.
- An inheritance on Strategy base class
- Rename load and save to load_local_state and save_local_state to improve understanding.
- Rename `build_graph` to `build_compute_plan`

fixes FL-640 

## Please check if the PR fulfills these requirements

- [ ] If the feature has an impact on the user experience, the [changelog](https://github.com/substra/substrafl/blob/main/CHANGELOG.md) has been updated
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)
- [ ] The commit message follows the [conventional commit](https://www.conventionalcommits.org/en/v1.0.0/) specification

### CODE REVIEW

**Comment by Owlfred** [comment]
End to end tests: :hourglass_flowing_sand: **STARTING**

File: substrafl/compute_plan_builder.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+        """Build the compute plan to be executed. All arguments are optional and will be feed within the
+        :func:`~substrafl.experiment.execute_experiment` function.
+
+        Args:
+            train_data_nodes (typing.List[TrainDataNode]): list of the train organizations
+            aggregation_node (typing.Optional[AggregationNode]): aggregation node, necessary for
+                centralized strategy, unused otherwise
+            evaluation_strategy (Optional[EvaluationStrategy]): evaluation strategy to follow for testing models.
+            num_rounds (int): Number of times to repeat the compute plan sub-graph (define in perform round).
+            clean_models (bool): Clean the intermediary models on the Substra platform. Set it to False
```
Comment (SdgJlbl) [inline substrafl/compute_plan_builder.py:49]: It does not default to True in the signature, is it a mishap?

File: substrafl/compute_plan_builder.py [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    ) -> Any:
+        """Build the compute plan to be executed. All arguments are optional and will be feed within the
+        :func:`~substrafl.experiment.execute_experiment` function.
+
+        Args:
+            train_data_nodes (typing.List[TrainDataNode]): list of the train organizations
+            aggregation_node (typing.Optional[AggregationNode]): aggregation node, necessary for
+                centralized strategy, unused otherwise
+            evaluation_strategy (Optional[EvaluationStrategy]): evaluation strategy to follow for testing models.
+            num_rounds (int): Number of times to repeat the compute plan sub-graph (define in perform round).
```
Comment (SdgJlbl) [inline substrafl/compute_plan_builder.py:48]: Do we want to add that it's only used in "recurring" graphs, and can be ignored else?

File: substrafl/strategies/strategy.py [hunk 3]
```diff
@@ -161,8 +161,8 @@ def build_graph(
             train_data_nodes (typing.List[TrainDataNode]): list of the train organizations
             aggregation_node (typing.Optional[AggregationNode]): aggregation node, necessary for
                 centralized strategy, unused otherwise
-            evaluation_strategy (Optional[EvaluationStrategy]): _description_
-            num_rounds (int): _description_
+            evaluation_strategy (Optional[EvaluationStrategy]): evaluation strategy to follow for testing models.
+            num_rounds (int): Number of times to repeat the compute plan sub-graph (define in perform round).
             clean_models (bool): Clean the intermediary models on the Substra platform. Set it to False
```
Comment (SdgJlbl) [inline substrafl/strategies/strategy.py:166]: Same as for the parent class, it does not default to True here 🤔

**Review by SdgJlbl — COMMENTED** [review]
Thanks for your work (and for splitting it in easy-to-review commits :) )
I think a few docstrings could be more explicit, and that a quite detailed changelog will be important to explain to the users why the change and how to use the new concept. 
Nice to see work advancing on that track 👍

File: substrafl/compute_plan_builder.py (hunk 1 above)
Comment (ThibaultFy) [reply substrafl/compute_plan_builder.py:49]: In fact the default is propagated from execute experiement, but this is not clean this way. I'll add the default here too thanks.

File: substrafl/compute_plan_builder.py (hunk 2 above)
Comment (ThibaultFy) [reply substrafl/compute_plan_builder.py:48]: Okay I'll developped this part. And I also added the optional default to none in execute experiment (after your comment). It was just blocking for the cp_metadata part, used in the download algo function. But it a good start IMO to pr

**Comment by ThibaultFy** [comment]
@SdgJlbl I got a circular import error due to the move to schemas in into strategy... I don't know why I'm only getting the error now, but the only solution I found is the one describe here (and on several other forum). Wht do you think about it ? https://adamj.eu/tech/2021/05/13/python-type-hints-how-to-fix-circular-imports/

**Comment by SdgJlbl** [comment]
> @SdgJlbl I got a circular import error due to the move to schemas in into strategy... I don't know why I'm only getting the error now, but the only solution I found is the one describe here (and on several other forum). Wht do you think about it ? https://adamj.eu/tech/2021/05/13/python-type-hints-how-to-fix-circular-imports/

Yes, if the issue is only linked to type checking, it's the way to go.

**Comment by Owlfred** [comment]
End to end tests: :hourglass_flowing_sand: **STARTING**

File: substrafl/algorithms/algo.py [hunk 4]
```diff
@@ -1,10 +1,15 @@
+from __future__ import annotations
```
Comment (SdgJlbl) [inline substrafl/algorithms/algo.py:1]: do we need both this and the TYPE_CHECKING trick?

File: substrafl/compute_plan_builder.py [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+
+    @abc.abstractmethod
+    def build_compute_plan(
+        self,
+        train_data_nodes: Optional[List[TrainDataNode]],
+        aggregation_node: Optional[List[AggregationNode]],
+        evaluation_strategy: Optional[EvaluationStrategy],
+        num_rounds: Optional[int],
+        clean_models: Optional[bool] = True,
+    ) -> Any:
```
Comment (SdgJlbl) [inline substrafl/compute_plan_builder.py:39]: Does it really returns Any? 
There is no docstring for the return, is it None?

File: substrafl/compute_plan_builder.py [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+                useful in recurring graphs, but can be ignored in other cases.
+            clean_models (bool): Clean the intermediary models on the Substra platform. Set it to False
+                if you want to download or re-use intermediary models. This causes the disk space to fill
+                quickly so should be set to True unless needed. Defaults to True.
+        """
+        pass
+
+    @abc.abstractmethod
+    def save_local_state(self, path: Path) -> Any:
+        """Executed at the end of each step of the computation graph to save
```
Comment (SdgJlbl) [inline substrafl/compute_plan_builder.py:58]: No docstring for the return

File: substrafl/strategies/strategy.py [hunk 7]
```diff
@@ -141,15 +141,15 @@ def perform_predict(
...
-    def build_graph(
+    def build_compute_plan(
         self,
         train_data_nodes: List[TrainDataNode],
         aggregation_node: Optional[List[AggregationNode]],
         evaluation_strategy: Optional[EvaluationStrategy],
         num_rounds: int,
-        clean_models: Optional[bool],
+        clean_models: Optional[bool] = True,
     ):
```
Comment (SdgJlbl) [inline substrafl/strategies/strategy.py:151]: -> None

**Review by SdgJlbl — APPROVED** [review]
A few comments related to type annotation, otherwise LGTM 👍 
Thanks for your work!

File: substrafl/algorithms/algo.py (hunk 4 above)
Comment (ThibaultFy) [reply substrafl/algorithms/algo.py:1]: Yes, it avoid changing the type hint "strategy name" to string:
`This isn’t necessary if we add from __future__ import annotations at the top of the file, which makes all annotations strings by default (see [PEP 563](https://www.python.org/dev/peps/pep-0563/#abstract)). (__future__.annotations will become the default in Python 3.11).`

File: substrafl/compute_plan_builder.py (hunk 5 above)
Comment (ThibaultFy) [reply substrafl/compute_plan_builder.py:39]: Nice catch... It returns None indeed

**Pull request merged**